pytng/pytng.pyx

```diff
@@ -93,25 +434,364 @@ cdef class MemoryWrapper:
+        #Flag to indicate particle dependent data. */
+        #define TNG_PARTICLE_DEPENDENT 2
+        cdef int  TNG_PARTICLE_DEPENDENT = 2
+
+        stat = tng_data_block_name_get(self._traj, block_id, name, TNG_MAX_STR_LEN)
+        if stat != TNG_SUCCESS:
+            raise Exception("cannot get block_name")
+
+        stat = tng_data_block_dependency_get(self._traj, block_id, &block_dependency)
+        if stat != TNG_SUCCESS:
+            raise Exception("cannot get block_dependency")
+        
+        if block_dependency.__and__(TNG_PARTICLE_DEPENDENT): # bitwise & due to enum defs
+            printf("reading particle data \n")
+            tng_num_particles_get(self._traj, n_atoms)
+            stat = tng_util_particle_data_next_frame_read(self._traj, block_id, &data, &datatype, step, frame_time)
+        else:
+            printf("reading NON particle data \n")
+            n_atoms[0] = 1 # still used for some allocs
+            stat = tng_util_non_particle_data_next_frame_read(self._traj, block_id, &data, &datatype, step, frame_time)
+
+        if stat == TNG_CRITICAL:
+            #raise Exception("critical data reading failure")
+            return TNG_CRITICAL
+
+        
+        stat = tng_data_block_num_values_per_frame_get(self._traj, block_id, n_values_per_frame)
+        if stat == TNG_CRITICAL:
+            #raise Exception("critical data reading failure")
+            return TNG_CRITICAL
+
+        # calling function is responsible for freeing value pointer
+        values[0] = <double*> realloc(values[0], sizeof(double)* n_values_per_frame[0] * n_atoms[0]) # renew to be right size
+        printf("realloc values array to be %ld  doubles and %ld bits long \n", n_values_per_frame[0] * n_atoms[0], n_values_per_frame[0] * n_atoms[0]*sizeof(double))
+
+        self.convert_to_double_arr(data, values[0], n_atoms[0], n_values_per_frame[0], datatype)
+
+        # free local data
+        free(data)
+
+        return TNG_SUCCESS
+
+    
+    cdef void convert_to_double_arr(self, void* source, double* to, const int n_atoms, const int n_vals, const char datatype):
+
+        # do we need to account for changes in the decl of double etc ie is this likely to be portable?.
+        # a lot of this is a bit redundant but could be used to differntiate casts to numpy arrays etc in the future
+
+        cdef int i, j
+
+
+        if datatype == TNG_FLOAT_DATA:
+            for i in range(n_atoms):
+                for j in range(n_vals):
+                    to[i*n_vals +j ] = (<float*>source)[i *n_vals +j] #NOTE do we explicitly need to use reinterpret_cast ??
+            #memcpy(to,  source, n_vals * sizeof(float) * n_atoms)
+
+        elif datatype == TNG_INT_DATA:
+            for i in range(n_atoms):
+                for j in range(n_vals):
+                    to[i*n_vals +j ] = (<int64_t*>source)[i *n_vals +j] # redundant but could be changed later
+            #memcpy(to, source, n_vals * sizeof(int64_t) * n_atoms)
```

Comment:
@richardjgowers  not sure if im calling this right  but these raw memcpy calls were not giving me anything that made sense.